File: frontend/src/services/attendance.ts
import api from './api';

export interface AttendanceRecord {
    id: number;
    employee_id: number;
    date: string;
    check_in?: string;
    check_out?: string;
    worked_hours?: number;
    extra_hours?: number;
    status: 'PRESENT' | 'ABSENT' | 'HALF_DAY' | 'LEAVE';
    created_at: string;
    updated_at: string;
    employee_name?: string;
    employee_code?: string;
    department_name?: string;
}

export const checkIn = async () => {
    const response = await api.post('/attendance/check-in');
    return response.data;
};

export const checkOut = async () => {
    const response = await api.post('/attendance/check-out');
    return response.data;
};

export const getTodayAttendance = async () => {
    const response = await api.get('/attendance/today');
    return response.data;
};

export const getMyAttendance = async (params?: { start_date?: string; end_date?: string; skip?: number; limit?: number }) => {
    const response = await api.get('/attendance/me', { params });
    return response.data;
};

export const getAdminAttendance = async (params?: {
    q?: string;
    employee_id?: number;
    department_id?: number;
    date?: string;
    start_date?: string;
    end_date?: string;
    status?: string;
    skip?: number;
    limit?: number;
}) => {
    const response = await api.get('/attendance', { params });
    return response.data;
};

export const getAttendanceById = async (id: number) => {
    const response = await api.get(`/attendance/${id}`);
    return response.data;
};

File: backend/app/models/attendance.py
from datetime import datetime, timezone
from sqlalchemy import Column, Integer, Date, DateTime, Float, ForeignKey, Enum, UniqueConstraint
from sqlalchemy.orm import relationship
import enum
from app.core.database import Base

class AttendanceStatusEnum(str, enum.Enum):
    PRESENT = "PRESENT"
    ABSENT = "ABSENT"
    HALF_DAY = "HALF_DAY"
    LEAVE = "LEAVE"

class Attendance(Base):
    __tablename__ = "attendances"

    id = Column(Integer, primary_key=True, index=True)
    employee_id = Column(Integer, ForeignKey("employees.id"), nullable=False, index=True)
    date = Column(Date, nullable=False, index=True)
    check_in = Column(DateTime, nullable=True)
    check_out = Column(DateTime, nullable=True)
    worked_hours = Column(Float, nullable=True)
    extra_hours = Column(Float, nullable=True)
    status = Column(Enum(AttendanceStatusEnum), default=AttendanceStatusEnum.PRESENT)
    
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    employee = relationship("Employee", back_populates="attendances")

    __table_args__ = (
        UniqueConstraint('employee_id', 'date', name='uq_attendance_employee_date'),
    )
